boards/BMS/Src/App/App_Eeprom.c

```diff
@@ -0,0 +1,207 @@
+    eeprom->read_page  = read_page;
+    eeprom->page_erase = page_erase;
+
+    return eeprom;
+}
+
+void App_Eeprom_Destroy(struct Eeprom *eeprom)
+{
+    free(eeprom);
+}
+
+EEPROM_StatusTypeDef
+    App_Eeprom_WriteFloats(struct Eeprom *eeprom, uint16_t page, uint8_t offset, float *input_data, uint8_t num_floats)
+{
+    // Check if data to write fits into single page
+    if (num_floats * sizeof(float) > PAGE_SIZE)
+    {
+        return EEPROM_SIZE_ERROR;
+    }
+    // for floats, offset should be aligned to multiple of 4
+    else if (offset % 4 != 0)
+    {
+        return EEPROM_ADDR_ERROR;
+    }
+
+    // convert float data to bytes array
+    uint8_t data_bytes[num_floats * sizeof(float)];
+
+    for (uint8_t i = 0; i < num_floats; i++)
+    {
+        convert_float_to_bytes(&data_bytes[i * sizeof(float)], input_data[i]);
+    }
+
+    // write bytes array to EEPROM
+    return eeprom->write_page(page, offset, data_bytes, (uint16_t)(num_floats * sizeof(float)));
+}
+
+EEPROM_StatusTypeDef
+    App_Eeprom_ReadFloats(struct Eeprom *eeprom, uint16_t page, uint8_t offset, float *output_data, uint8_t num_floats)
+{
+    // Check if data to read fits into single page
+    if (num_floats * sizeof(float) + offset > PAGE_SIZE)
+    {
+        return EEPROM_SIZE_ERROR;
+    }
+
+    uint8_t data[num_floats * sizeof(float)];
+    for (int i = 0; i < num_floats; i++)
+    {
+        data[i] = 0;
+    }
+    uint8_t read_status;
+
+    read_status = eeprom->read_page(page, offset, data, (uint16_t)(num_floats * sizeof(float)));
+
+    for (uint8_t i = 0; i < num_floats; i++)
+    {
+        output_data[i] = convert_bytes_to_float(&data[i * sizeof(float)]);
+    }
+
+    return read_status;
+}
+
+EEPROM_StatusTypeDef App_Eeprom_PageErase(struct Eeprom *eeprom, uint16_t page)
+{
+    return eeprom->page_erase(page);
+}
+
+EEPROM_StatusTypeDef App_Eeprom_WriteAddress(struct Eeprom *eeprom, uint16_t page, uint16_t address)
+{
+    uint16_t num_bytes = SAVED_COPIES * 2; // saving 3 copies of address, each 2 bytes
+    uint8_t  byte_array[num_bytes];
+    uint8_t  offset = 0;
+
+    for (uint8_t i = 0; i < SAVED_COPIES; i++)
+    {
+        convert_short_to_bytes(&byte_array[i * 2], address);
+    }
+
+    return eeprom->write_page(page, offset, byte_array, num_bytes);
+}
+
+EEPROM_StatusTypeDef App_Eeprom_ReadAddress(struct Eeprom *eeprom, uint16_t page, uint16_t *address)
+{
+    uint16_t num_bytes = SAVED_COPIES * 2; // saving 3 copies of address, each 2 bytes
+    uint8_t  byte_array[num_bytes];
+    uint8_t  offset = 0;
+    uint8_t  read_status;
+
+    read_status = eeprom->read_page(page, offset, byte_array, num_bytes);
+
+    uint16_t short_array[SAVED_COPIES];
+
+    for (uint8_t i = 0; i < SAVED_COPIES; i++)
+    {
+        short_array[i] = convert_bytes_to_short(&byte_array[i * 2]);
+    }
+
+    // TODO: Implement error checking using 3 copies of address stored in short_array
+    *address = short_array[0];
+
```

Comment:
I think it makes sense to do all the voting in the same file: Since it looks like you're doing voting in the upcoming SOC file for the floats, might as well do the address voting there too. So consider making this a "ReadAddresses" function like the "ReadFloats" above.